### ISSUE TITLE
Adapters should be passed config-in-effect, not config-as-set

### ISSUE THREAD

**Issue by duizendnegen** [LevelbossMike/ember-deploy#29]
Currently, with configuring gzip extensions, both ember-deploy and (in my case) ember-deploy-azure have to check `config.assets.gzipExtensions ? config.assets.gzipExtensions : ['js', 'css', 'svg']`.

It would be much cleaner and less prone to error (e.g. for when ember-deploy changes something in its default settings and an old ember-deploy-addon is not updated) to have ember-deploy augment the configuration file with default settings and passing that on to underlying deploy-addons. Possibly ideally in the configuration reader.

As I don't have a clear picture of which implications it has for existing infrastructure, just thought I'd open an issue to get some feedback first.

**Comment by LevelbossMike** [comment]
yeah. the config object (i.e.) configuration reader should be responsible to know of the sane defaults. totally agree on that.

**Comment by jamesfid** [comment]
### Approach
Here's a rough idea of what could happen to the `configuration-reader` to default the values.  This is simplistic, but I wanted to further the discussion on how this should look:
```js
var path       = require('path');

// Set a value on the object if it is not already set.
// => Will not override any configuration values from deploy.json, but will set them if they are not present.
// => Code reference: http://stackoverflow.com/questions/13719593/how-to-set-object-property-of-object-property-of-given-its-string-name-in-ja
var defaultValue = function(obj, prop, value) {
    if (typeof prop === "string")
        prop = prop.split(".");

    if (prop.length > 1) {
        var e = prop.shift();
        defaultValue(obj[e] =
                 Object.prototype.toString.call(obj[e]) === "[object Object]"
                 ? obj[e]
                 : {},
               prop,
               value);
    } else
      prop = prop[0];
      if (typeof obj[prop] === 'undefined')
        obj[prop] = value;
}

function ConfigurationReader(options) {
  var envConfig, root = process.cwd();

  this._options      = options || {};
  this._environment  = this._options.environment || 'development';
  this._deployConfig = this._options.configFile || 'deploy.json';
  this._config       = require(path.join(root, this._deployConfig));
  this._envConfig    = this._config[this._environment];

  defaultValue(this._envConfig,'assets.type','s3');
  defaultValue(this._envConfig,'assets.gzip',true);
  defaultValue(this._envConfig,'assets.gzipExtensions',['js', 'css', 'svg']);
  defaultValue(this._envConfig,'store.type','redis');
  defaultValue(this._envConfig,'buildEnv','production');

  this.store    = this._envConfig.store;
  this.assets   = this._envConfig.assets;
  this.buildEnv = this._envConfig.buildEnv;
}

module.exports = ConfigurationReader;
```
[gist](https://gist.github.com/jamesfid/acf01c50531246107ebe)

### Notes
This will break two of the tests that are expecting the configuration to **not** have certain keys, as they are set via an 'or' instead of in the `configuration-reader`.  If the approach we take is something like the above, the test version of `deploy.json` will need to be updated to include the newly present keys.

We could theoretically make this a CoreObject extension, but I'm not sure if that provides any additional benefits at this point.

### Moving Forward
@duizendnegen, @LevelbossMike -- any feedback on this approach?  If we can come up with how we want it to be handled, I can write it and fix the tests, remove defaulting in adapters, etc.

**Comment by duizendnegen** [comment]
That would probably be the best approach - as I see it now also with the least effort & best future possibilities to extend upon.

**Comment by lukemelia** [comment]
@duizendnegen please take a look at #37 -- I think it tackles most of what you proposed here, albeit a little differently. Would love to hear your thoughts.

**Comment by LevelbossMike** [comment]
I have just merged #37. I think this fixes the issue and does this in a clean way.

**Comment by jamesfid** [comment]
#37 is great - glad to see we got something merged in!
